src/paulie/application/pauli_compiler_optimal.py

```diff
@@ -0,0 +1,612 @@
+                return []
+            r = len(ui_bi)
+            i = r - 1
+            G: list[PauliString] = [self.extend_pair(ui_bi[-1][0], ui_bi[-1][1])]
+            H: list[PauliString] = []
+            helper_used_for_i: dict[int, int] = {}
+            while i >= 1:
+                Ui, Bi = ui_bi[i]
+                P_left, P_right = self._rest_full_after(ui_bi, i, H)
+                if _multiply(P_left, Ui).is_identity():
+                    cnt = helper_used_for_i.get(i, 0)
+                    if cnt >= 1:
+                        G.append(self.extend_pair(Ui, Bi))
+                        i -= 1
+                        continue
+                    A1, A2 = self._choose_a1_a2(Ui)
+                    H = [A1, A2]
+                    helper_used_for_i[i] = cnt + 1
+                    G.append(self.extend_left(A1))
+                    G.append(self.extend_left(A2))
+                    continue
+                current = self.extend_pair(Ui, Bi)
+                rest_full = _tensor(P_left, P_right)
+                if _commutes(current, rest_full):
+                    cnt = helper_used_for_i.get(i, 0)
+                    if cnt >= 1:
+                        G.append(self.extend_pair(Ui, Bi))
+                        i -= 1
+                        continue
+                    A_prime = self._choose_aprime(Ui, P_left)
+                    H = [A_prime]
+                    helper_used_for_i[i] = cnt + 1
+                    G.append(self.extend_left(A_prime))
+                    continue
+                G.append(current)
+                i -= 1
+            return G
+        return []
+
+
+def _key(p: PauliString) -> str:
+    """Convert PauliString to string"""
+    return str(p)
+
+def left_map_over_a(V_from: PauliString, V_to: PauliString, A: list[PauliString]
+) -> list[PauliString]:
+    """left map over a"""
+    if _key(V_from) == _key(V_to):
+        return []
+    start_k = _key(V_from)
+    goal_k = _key(V_to)
+    q: deque[PauliString] = deque([V_from])
+    parent: dict[str, tuple[str, str, PauliString]] = {}
+    seen = {start_k}
+    while q:
+        cur = q.popleft()
+        cur_k = _key(cur)
+        if cur_k == goal_k:
+            seq: list[PauliString] = []
+            while cur_k != start_k:
+                prev_k, _, a_used = parent[cur_k]
+                seq.append(a_used)
+                cur_k = prev_k
+            seq.reverse()
+            return seq
+        for a in A:
+            if _commutes(a, cur):
+                continue
+            nxt = _multiply(a, cur)
+            nk = _key(nxt)
+            if nk in seen:
+                continue
+            seen.add(nk)
+            parent[nk] = (cur_k, nk, a)
+            q.append(nxt)
+    raise RuntimeError("Left map BFS failed.")
+
+
+@dataclass
+class PauliCompilerConfig:
+    """Config of PauliCompiler"""
+    k_left: int
+    n_total: int
+    fallback_depth: int = 8
+    fallback_nodes: int = 200000
+
+class OptimalPauliCompiler:
```

Comment:
What is this optimal with respect to? Should we just caull this PauliCompiler? We probably should as the PauliCompilerConfig dataclass above corresponds to it, so it's confusing to have two different names in this way.